File: examples/kafka/go/producer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/IBM/sarama"
)

// TaskEvent represents a task-related event
type TaskEvent struct {
	EventType string    `json:"event_type"`
	TaskID    string    `json:"task_id"`
	Title     string    `json:"title"`
	UserID    string    `json:"user_id"`
	Status    string    `json:"status"`
	Timestamp time.Time `json:"timestamp"`
}

func main() {
	// Kafka configuration
	config := sarama.NewConfig()
	config.Version = sarama.V3_6_0_0
	config.Producer.Return.Successes = true
	config.Producer.RequiredAcks = sarama.WaitForAll
	config.Producer.Retry.Max = 5
	config.Producer.Compression = sarama.CompressionLZ4

	// SASL Authentication
	config.Net.SASL.Enable = true
	config.Net.SASL.Mechanism = sarama.SASLTypeSCRAMSHA256
	config.Net.SASL.User = getEnv("KAFKA_USERNAME", "kafka-user")
	config.Net.SASL.Password = getEnv("KAFKA_PASSWORD", "kafka-secret-password")

	// Kafka brokers
	brokers := []string{getEnv("KAFKA_BROKERS", "kafka.kafka.svc.cluster.local:9092")}
	topic := getEnv("KAFKA_TOPIC_TASKS", "tasks")

	// Create producer
	producer, err := sarama.NewSyncProducer(brokers, config)
	if err != nil {
		log.Fatalf("Failed to create producer: %v", err)
	}
	defer producer.Close()

	log.Printf("✅ Connected to Kafka brokers: %v", brokers)
	log.Printf("📤 Publishing to topic: %s", topic)

	// Example 1: Task Created Event
	taskCreated := TaskEvent{
		EventType: "task.created",
		TaskID:    "task-123",
		Title:     "Implement user authentication",
		UserID:    "user-456",
		Status:    "pending",
		Timestamp: time.Now(),
	}

	if err := publishEvent(producer, topic, taskCreated); err != nil {
		log.Printf("❌ Failed to publish task.created: %v", err)
	}

	// Example 2: Task Updated Event
	time.Sleep(1 * time.Second)
	taskUpdated := TaskEvent{
		EventType: "task.updated",
		TaskID:    "task-123",
		Title:     "Implement user authentication",
		UserID:    "user-456",
		Status:    "in_progress",
		Timestamp: time.Now(),
	}

	if err := publishEvent(producer, topic, taskUpdated); err != nil {
		log.Printf("❌ Failed to publish task.updated: %v", err)
	}

	// Example 3: Task Completed Event
	time.Sleep(1 * time.Second)
	taskCompleted := TaskEvent{
		EventType: "task.completed",
		TaskID:    "task-123",
		Title:     "Implement user authentication",
		UserID:    "user-456",
		Status:    "completed",
		Timestamp: time.Now(),
	}

	if err := publishEvent(producer, topic, taskCompleted); err != nil {
		log.Printf("❌ Failed to publish task.completed: %v", err)
	}

	log.Println("🎉 All events published successfully!")
}

// publishEvent publishes an event to Kafka
func publishEvent(producer sarama.SyncProducer, topic string, event TaskEvent) error {
	// Serialize event to JSON
	payload, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	// Create Kafka message
	msg := &sarama.ProducerMessage{
		Topic: topic,
		Key:   sarama.StringEncoder(event.TaskID), // Use TaskID as partition key
		Value: sarama.ByteEncoder(payload),
		Headers: []sarama.RecordHeader{
			{
				Key:   []byte("event_type"),
				Value: []byte(event.EventType),
			},
		},
	}

	// Send message
	partition, offset, err := producer.SendMessage(msg)
	if err != nil {
		return fmt.Errorf("failed to send message: %w", err)
	}

	log.Printf("✅ Published %s to partition %d at offset %d", event.EventType, partition, offset)
	return nil
}

// getEnv gets environment variable with fallback
func getEnv(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}

File: examples/kafka/nodejs/producer.js
const { Kafka, CompressionTypes } = require('kafkajs');

// Kafka configuration
const kafka = new Kafka({
  clientId: 'user-service-producer',
  brokers: [process.env.KAFKA_BROKERS || 'kafka.kafka.svc.cluster.local:9092'],
  sasl: {
    mechanism: 'scram-sha-256',
    username: process.env.KAFKA_USERNAME || 'kafka-user',
    password: process.env.KAFKA_PASSWORD
  },
  retry: {
    initialRetryTime: 300,
    retries: 10
  }
});

const producer = kafka.producer({
  compression: CompressionTypes.LZ4,
  idempotent: true,
  maxInFlightRequests: 5,
  transactionalId: 'user-service-transactional-producer'
});

// User event types
const UserEventType = {
  CREATED: 'user.created',
  UPDATED: 'user.updated',
  DELETED: 'user.deleted',
  LOGIN: 'user.login',
  LOGOUT: 'user.logout'
};

/**
 * Publish user event to Kafka
 */
async function publishUserEvent(eventType, userData) {
  const event = {
    event_type: eventType,
    user_id: userData.userId,
    email: userData.email,
    name: userData.name,
    timestamp: new Date().toISOString(),
    metadata: userData.metadata || {}
  };

  try {
    const result = await producer.send({
      topic: process.env.KAFKA_TOPIC_USERS || 'users',
      messages: [
        {
          key: userData.userId, // Partition by user ID
          value: JSON.stringify(event),
          headers: {
            'event_type': eventType,
            'service': 'user-service',
            'version': '1.0'
          }
        }
      ]
    });

    console.log(`✅ Published ${eventType}:`, {
      userId: userData.userId,
      partition: result[0].partition,
      offset: result[0].offset
    });

    return result;
  } catch (error) {
    console.error(`❌ Failed to publish ${eventType}:`, error);
    throw error;
  }
}

/**
 * Main function
 */
async function main() {
  try {
    // Connect producer
    await producer.connect();
    console.log('✅ Producer connected to Kafka');

    // Example 1: User Registration
    await publishUserEvent(UserEventType.CREATED, {
      userId: 'user-789',
      email: 'newuser@example.com',
      name: 'John Doe',
      metadata: {
        source: 'web',
        ip: '192.168.1.100'
      }
    });

    // Wait a bit
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Example 2: User Login
    await publishUserEvent(UserEventType.LOGIN, {
      userId: 'user-789',
      email: 'newuser@example.com',
      name: 'John Doe',
      metadata: {
        device: 'web-browser',
        ip: '192.168.1.100',
        location: 'Vietnam'
      }
    });

    // Wait a bit
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Example 3: User Profile Update
    await publishUserEvent(UserEventType.UPDATED, {
      userId: 'user-789',
      email: 'newuser@example.com',
      name: 'John Doe Updated',
      metadata: {
        fields_updated: ['name', 'avatar'],
        updated_by: 'user-789'
      }
    });

    // Wait a bit
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Example 4: User Logout
    await publishUserEvent(UserEventType.LOGOUT, {
      userId: 'user-789',
      email: 'newuser@example.com',
      name: 'John Doe Updated',
      metadata: {
        session_duration: 3600,
        device: 'web-browser'
      }
    });

    console.log('🎉 All events published successfully!');

  } catch (error) {
    console.error('❌ Error in main:', error);
    process.exit(1);
  } finally {
    // Disconnect
    await producer.disconnect();
    console.log('👋 Producer disconnected');
  }
}

// Handle graceful shutdown
process.on('SIGINT', async () => {
  console.log('\n⚠️  SIGINT received, shutting down...');
  await producer.disconnect();
  process.exit(0);
});

process.on('SIGTERM', async () => {
  console.log('\n⚠️  SIGTERM received, shutting down...');
  await producer.disconnect();
  process.exit(0);
});

// Run if executed directly
if (require.main === module) {
  main().catch(console.error);
}

// Export for use in other modules
module.exports = {
  producer,
  publishUserEvent,
  UserEventType
};
